engine/execution/computation/metrics/provider.go

```diff
@@ -0,0 +1,120 @@
+package metrics
+
+import (
+	"sync"
+
+	"github.com/rs/zerolog"
+)
+
+// provider is responsible for providing the metrics for the rpc endpoint
+// it has a circular buffer of metrics for the last N finalized and executed blocks.
+type provider struct {
+	log zerolog.Logger
+
+	mu sync.RWMutex
+
+	bufferSize               uint
+	bufferIndex              uint
+	blockHeightAtBufferIndex uint64
+
+	buffer [][]TransactionExecutionMetrics
+}
+
+func newProvider(
+	log zerolog.Logger,
+	bufferSize uint,
+	blockHeightAtBufferIndex uint64,
+) *provider {
+	if bufferSize == 0 {
+		panic("buffer size must be greater than zero")
+	}
+
+	return &provider{
+		log:                      log,
+		bufferSize:               bufferSize,
+		blockHeightAtBufferIndex: blockHeightAtBufferIndex,
+		bufferIndex:              0,
+		buffer:                   make([][]TransactionExecutionMetrics, bufferSize),
+	}
+}
+
+func (p *provider) Push(
```

Comment:
```suggestion
// Push buffers the metrics for the given height.
// The call should ensure height are called in strictly increasing order, otherwise 
// metrics for the skipped height will not buffered.
func (p *provider) Push(
```